### ISSUE TITLE
about voting_evalute_cls.py

### ISSUE THREAD

**Issue by zouyanmei** [tiangexiang/CurveNet#17]
Thank you for sharing. I tried to reproduce the voting_evalute_cls.py of RS-CNN, but  I did not get a satisfactory result about your pre-train models. So if it is convenient for you, I hope you can provide the code you use. Thinks again.

**Comment by zouyanmei** [comment]
If I modify it directly from the RS-CNN, there is a problem with different PyTorch versions, and some functions need to be replaced. I have no idea about it, so I watch to know how you resolve this problem. Thanks.

**Comment by tiangexiang** [comment]
Hi, we only modify the voting script to work under our environment and data loader with no other substantial changes therefore we didn't keep particular records of the script itself. 

One possible reason is that the FPS in our repo has a fixed starting point, while FPS used in the voting script may rely on complete randomness, please refer to their version of implementation for this.

**Comment by zouyanmei** [comment]
Thanks.The code I edited is like this,`
import os
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
from data import ModelNet40
from models.curvenet_cls import CurveNet
import numpy as np
from torch.utils.data import DataLoader
from util import IOStream, load_cfg_from_cfg_file, merge_cfg_from_list, cal_loss
import sklearn.metrics as metrics
import random
from tqdm import tqdm
from torchvision import transforms


def get_parser():
    parser = argparse.ArgumentParser(description='3D Object Classification')
    parser.add_argument('--exp_name', type=str, default='curvenet_cls', metavar='N', help='Name of the experiment')
    parser.add_argument('--num_points', type=int, default=1024, help='num of points to use')
    parser.add_argument('--no_cuda', type=bool, default=False, help='enables CUDA training')
    parser.add_argument('--test_batch_size', type=int, default=16, metavar='batch_size', help='Size of batch)')
    parser.add_argument('--manual_seed', type=int, default=1, help='random seed')
    parser.add_argument('--workers', type=int, default=6, help="workers")
    args = parser.parse_args()
    return args


def _init_():
    if not os.path.exists('../checkpoints'):
        os.makedirs('../checkpoints')
    if not os.path.exists('../checkpoints/'+args.exp_name):
        os.makedirs('../checkpoints/'+args.exp_name)

    # backup the running files:
    os.system('cp eval_voting.py ../checkpoints/' + '' + args.exp_name + '/' + 'eval_voting.py.backup')


class PointcloudScale(object):  # input random scaling
    def __init__(self, scale_low=2. / 3., scale_high=3. / 2.):
        self.scale_low = scale_low
        self.scale_high = scale_high

    def __call__(self, pc):
        bsize = pc.size()[0]
        for i in range(bsize):
            xyz = np.random.uniform(low=self.scale_low, high=self.scale_high, size=[3])
            scales = torch.from_numpy(xyz).float().cuda()
            pc[i, :, 0:3] = torch.mul(pc[i, :, 0:3], scales)
        return pc


def test(args, io):
    test_loader = DataLoader(ModelNet40(partition='test', num_points=args.num_points), num_workers=args.workers,
                             batch_size=args.test_batch_size, shuffle=False, drop_last=False)

    device = torch.device("cuda" if args.cuda else "cpu")
    NUM_PEPEAT = 300
    NUM_VOTE = 10

   #Try to load models
    model = CurveNet().to(device)
    model = nn.DataParallel(model)
    model.load_state_dict(torch.load( '../checkpoints/%s/models/model.t7' % args.exp_name))
    model = model.eval()
    best_acc = 0
    best_class =0 

    pointscale = PointcloudScale(scale_low=0.8, scale_high=1.18)  # set the range of scaling

    for i in range(NUM_PEPEAT):
        test_true = []
        test_pred = []

        for j, dataload in tqdm(enumerate(test_loader,0), total=len(test_loader), smoothing=0.9):
            data, label = dataload
            data, label = data.to(device), label.to(device).squeeze()
            data = data.permute(0, 2, 1)
            pred = 0
            for v in range(NUM_VOTE):
                new_data = data
                batch_size = data.size()[0]
                if v > 0:
                    new_data.data = pointscale(new_data.data)
                with torch.no_grad():
                    pred += F.softmax(model(new_data), dim=1)  # sum 10 preds
            pred /= NUM_VOTE   # avg the preds!
            label = label.view(-1)
            pred_choice = pred.max(dim=1)[1]
            test_true.append(label.cpu().numpy())
            test_pred.append(pred_choice.detach().cpu().numpy())
        test_true = np.concatenate(test_true)
        test_pred = np.concatenate(test_pred)
        test_acc = metrics.accuracy_score(test_true, test_pred)
        avg_per_class_acc = metrics.balanced_accuracy_score(test_true, test_pred)
        if test_acc > best_acc:
            best_acc = test_acc
            best_class = avg_per_class_acc
        outstr = 'Voting %d, test acc: %.6f, test avg acc: %.6f' % (i, test_acc * 100, avg_per_class_acc*100)
        io.cprint(outstr)

    final_outstr = 'Final voting test acc: %.6f, test class acc:' % (best_acc * 100, best_class*100)
    io.cprint(final_outstr)


if __name__ == "__main__":
    args = get_parser()
    _init_()

    io = IOStream('../checkpoints/'+ args.exp_name+'/%s_voting.log' % args.exp_name)
    io.cprint(str(args))

    if args.manual_seed is not None:
        random.seed(args.manual_seed)
        np.random.seed(args.manual_seed)
        torch.manual_seed(args.manual_seed)

    args.cuda = not args.no_cuda and torch.cuda.is_available()

    if args.cuda:
        io.cprint('Using GPU')
        if args.manual_seed is not None:
            torch.cuda.manual_seed(args.manual_seed)
            torch.cuda.manual_seed_all(args.manual_seed)
    else:
        io.cprint('Using CPU')

    test(args, io)
`. then, My test result is 93.88%。

**Comment by zouyanmei** [comment]
Based on the outstanding model CurveNet you proposed, I added the modules I proposed to get a small improvement. The final result is stable at 93.96% acc and 91.27%macc. the model is [https://drive.google.com/drive/folders/1AJLkP128JaWIP9RlK7UxahNBYqkJ6k5K](curvenet_new).

**Comment by tiangexiang** [comment]
Hi, for the voting, it seems that you forgot to re-sample 1200 points (using FPS) from the raw point cloud and randomly pick a new set of 1024 points as the input for the network (same to RSCNN), this could be a reason for the low accuracy.

Very impressed to see that you have reached 93.96% for ModelNet40!! To the best of my knowledge, your result could be the new SOTA on this benchmark. If your proposed modules are novel enough, or there is still space for further improvement, you definitely could composite a good research paper on this!
